contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,302 @@
+# Scheduling Policy
+
+_Status: Draft_
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @tallclair, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work.
+-  How NodeSelector works.
+-  How node / pod affinity / anti-affinity rules work.
+-  How several schedulers can be used within a single cluster.
+-  How priority classes work.
+-  How to set defaults in Kubernetes.
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namespaces, it allows administrators to:
+
+-  Restrict execution for specific applications (which are namespace scoped) into certain nodes
+-  Create policies that prevent users from even attempting to schedule workloads onto masters to maximize security
+-  require that a pods under a namespace run on dedicated nodes
+-  Restrict usage of some `PriorityClass`
+-  Restrict usage to a specific set of schedulers.
+-  enforcing pod affinity or anti-affinity rules on some particular namespace.
+
+Also Mandatory values must also be enforced by scheduling policies in case of:
+
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters (will also be handled later by the [RuntimeClass]() KEP)
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+### syntaxic standards
+
+### empty match and unset fields:
+
+We follow Kubernetes conventions, an empty field matches everything an unset one matches nothing
```

Comment:
I think this is backwards. Unset should match everything, empty should match the empty value.

Care will need to be taken when the unset value and empty value are the same. For example, consider:
```
SchedulerName []string
```
If I have `SchedulerName = nil` (or equivalently `[]string{}`) , then that MUST match any scheduler name. If I want to explicitly match the unset ScehdulerName, I would need to set `SchedulerName = []string{''}`.

Unfortunately I think this will need to be considered on a case-by-case basis, as the behavior depends on both the governed field type (the podspec field), and the policy field type.

Reply:
Any API that distinguishes between a nil slice and an empty slice is likely to be broken.  The conversion semantics do not always retain the difference, and I strongly urge you to not distinguish.